src/Utils/isAssignableTo.ts

```diff
@@ -305,3 +427,13 @@ export function isAssignableTo(
 
     return false;
 }
+
+function setInferredType(key: string, source: BaseType, inferMap: Map<string, BaseType>): void {
+    const infer = inferMap.get(key);
+
+    if (infer === undefined) {
+        inferMap.set(key, source);
+    } else {
+        inferMap.set(key, new UnionType([infer, source]));
```

Comment:
check if `infer` is a union type and append to it instead of creating N levels of `UnionType`.